spark/src/main/scala/org/apache/spark/sql/delta/schema/ImplicitMetadataOperation.scala

```diff
@@ -227,6 +226,108 @@ object ImplicitMetadataOperation {
+      spark: SparkSession,
+      path: Seq[String],
+      metadata: Metadata,
+      currentDt: DataType,
+      updateDt: DataType): Unit = {
+    val dependentConstraints =
+      Constraints.findDependentConstraints(spark, path, metadata)
+    if (dependentConstraints.nonEmpty) {
+      throw DeltaErrors.constraintDataTypeMismatch(
+        path,
+        currentDt,
+        updateDt,
+        dependentConstraints
+      )
+    }
+  }
+
+  /**
+   * Check whether there are dependant generated columns for
+   * the provided `currentDt`; if so, throw an error indicating
+   * the generated columns data type mismatch.
+   *
+   * @param spark the spark session used.
+   * @param path the full column path.
+   * @param protocol the protocol used.
+   * @param metadata the metadata used for checking dependant generated columns.
+   * @param currentDt the current data type.
+   * @param updateDt the updated data type.
+   */
+  private def checkDependentGeneratedColumns(
+      spark: SparkSession,
+      path: Seq[String],
+      protocol: Protocol,
+      metadata: Metadata,
+      currentDt: DataType,
+      updateDt: DataType): Unit = {
+    val dependentGeneratedColumns = SchemaUtils.findDependentGeneratedColumns(
+      spark, path, protocol, metadata.schema)
+    if (dependentGeneratedColumns.nonEmpty) {
+      throw DeltaErrors.generatedColumnsDataTypeMismatch(
+        path,
+        currentDt,
+        updateDt,
+        dependentGeneratedColumns
+      )
+    }
+  }
+
+  /**
+   * If the provided data types are `StructType`, check whether the nested/inner fields
+   * of the struct are being referenced by CHECK constraints or generated columns.
+   * It is worth noting that we only care about the inner fields of `StructType`,
+   * for all the other types, e.g., `MapType` or `ArrayType`, we will directly check
+   * the dependant constraints and generated columns without inspecting the inner fields.
+   *
+   * @param spark the spark session used.
+   * @param protocol the protocol used.
+   * @param path the full column path.
+   * @param from the current data type.
+   * @param to the updated data type.
+   * @param metadata the metadata used for checking constraints and generated columns.
+   */
+  private def checkReferencedByCheckConstraintsOrGeneratedColumns(
+      spark: SparkSession,
+      protocol: Protocol,
+      path: Seq[String],
+      from: DataType,
+      to: DataType,
+      metadata: Metadata): Unit = (from, to) match {
+    // we only check for (nested) struct fields
+    case (StructType(fromFields), StructType(toFields)) =>
+      fromFields.zip(toFields).foreach { case (fromField, toField) =>
+        checkReferencedByCheckConstraintsOrGeneratedColumns(
+          spark = spark,
+          protocol = protocol,
+          path = path :+ fromField.name,
+          from = fromField.dataType,
+          to = toField.dataType,
+          metadata = metadata
+        )
+      }
+    case (fromDataType, toDataType) =>
```

Comment:
@johanl-db thanks for the review!

after investigating a bit, it seems that we can not directly access the field of a nested `struct` inside a `array`, e.g.,

```sql
create table t (a struct<b: array<struct<x: byte, y: byte>>>) using delta;

-- this will fail due to - "[FIELD_NOT_FOUND] No such struct field `element` in `x`, `y`."
-- same for generated columns.
alter table t add constraint ck check (hash(a.b.element.x) > 0);
```

there could be other way to achieve this, e.g., `alter table t add constraint ck check (hash(a.b[0].x) > 0)`, but this will cause other issues - even if we check for the generated columns/constraints, the path will not be consistent as `a.b[0].x` since we automatically append the `"element"` in [SchemaMergingUtils.transformColumns](https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/schema/SchemaMergingUtils.scala#L311), which will lead to the checks for dependant constraints/generated columns being bypassed even if we've checked the field for the two properties, i.e., `a.b.element.x` (actual) is not consistent with `a.b[0].x` (expected).

I think we could probably block `ArrayType` and `MapType` (due to the appended `"key"` and `"value"`) for now, and only relax the check for `StructType`.

Reply:
Agree, it's better to keep what you have currently and keep being too strict for map and arrays - it's not worth the effort and risk of bugs